lib/agentx.c

```diff
@@ -259,6 +262,16 @@ void smux_init(struct thread_master *tm)
 	install_element(CONFIG_NODE, &no_agentx_cmd);
 }
 
+void smux_agentx_enable(void)
+{
+	if (!agentx_enabled) {
+		init_snmp(FRR_SMUX_NAME);
+		events = list_new();
+		agentx_events_update();
+		agentx_enabled = 1;
+	}
+}
```

Comment:
I don't see where this is used with the hook?

Reply:
Hi @qlyoung, 

Originally, I posted this change as part of a bigger PR: Add SNMP support to LDP.
Here is the 'Add SNMP support to LDP' PR: https://github.com/FRRouting/frr/pull/7944

A comment on PR 7944 asked that I post this agentx_enabled hook as a separate PR.
So I created this PR for just the agentx_enabled hook.

I can understand if it's easier to review the agentx_enabled hook with the LDP SNMP PR.
If that's the case, just let me know & I can close this PR and continue with just the 1 PR (7944).

Thanks, Karen

Reply:
Nah the PR is fine, I'm just wondering what that function I commented on is doing here, since I don't see it either using or being used by the hook.

Reply:
Hi @qlyoung,

The short answer is: the LDP SNMP PR calls smux_agentx_enable().

I'll include the more complete answer below.

If I can answer other questions, just let me know...  

Thanks, Karen


- LDP runs as 3 processes.
   - LDP main is a client of FRR.
   - The other 2 LDP processes (lde, ldpe) are not FRR clients.
   - These other 2 LDP processes contain the data that the MIBs expose.  In the LDP SNMP PR, these other 2 LDP processes register MIB callbacks with the agent.
- LDP main registers for the agentx_enabled hook.
- When the agentx_enabled hook is called, LDP main sends an IMSG_AGENTX_ENABLED message to the other 2 LDP processes.
- When the other 2 LDP processes (lde, ldpe) receive they IMSG_AGENTX_ENABLED, they perform the following:
   - load the LDP SNMP module.
   - Initialize with agent by calling:
       - smux_init(tm);
       - smux_agentx_enable();
       - REGISTER_MIB(...)